### PR TITLE
[Serve] [Dashboard] Add end times and DELETED state for endpoints

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
This PR stores the history of deleted deployments and their end times in the Serve checkpoint and shows this information in the cluster snapshot.  

Currently the list of deleted deployments can grow without bound, so in a future PR we will need to add some way of limiting the size.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by architkulkarni** [comment]
cc @kombuchafox

File: python/ray/serve/backend_state.py
Comment (simon-mo) [inline python/ray/serve/backend_state.py:544]: let's at least bound this with ordereddict? https://docs.python.org/3/library/collections.html#collections.OrderedDict
```python
d = OrderedDict()
if len(d) >= 100:
    d.popitem(last=False)
d[k] = v
```

**Comment by simon-mo** [comment]
@jiaodong can you help review this PR as it relates to BackendState?

File: python/ray/serve/backend_state.py
Comment (jiaodong) [reply python/ray/serve/backend_state.py:544]: Can we call it `deleted_backend_metadata` instead since it's exactly the same type and content as `_backend_metadata` with same key value for assignment, so we don't have to introduce another term like "deployment" ?

Also, what happens if the same backend tag (for example, "/endpoint") got deployed and deleted multiple times, do we intend to always keep the last one or rather keep history of deployments for the same endpoint overtime ? From my local script 

```
import ray
from ray import serve

ray.init()
serve.start()

@serve.deployment(num_replicas=1)
class ConstructorFailureDeployment:
    def __init__(self):
        raise RuntimeError("Intentionally throwing ")

    async def serve(self, request):
        return "hi"

ConstructorFailureDeployment.deploy()
```


Our backend state thinks backend tag is "ConstructorFailureDeployment" which seems to just be the class name without any randomized suffix, thus the chance of reusing the same backend tag could be high.

**Review by jiaodong — COMMENTED** [review]
Just adding `Dict[BackendTag, BackendInfo]` in checkpoint looks good to me since it's same type and content as `self._backend_metadata` that we're already checkpointing and not targeting to remove, so it didn't introduce any new incompatibility.

File: python/ray/serve/backend_state.py
Comment (architkulkarni) [reply python/ray/serve/backend_state.py:544]: `deleted_backend_metadata` name change sounds good, I'll change it.  

Regarding repeated deletes and deploys of a given deployment, just discussed this on Slack with @kombuchafox and for now the current behavior (always keep the last one) is okay, the consumer of the snapshot will keep track of the historical data.

**Comment by architkulkarni** [comment]
Windows builds are stalled on master

**Comment by architkulkarni** [comment]
@alanwguo @ijrsvt @joeybai @wuisawesome could one of you please stamp this PR?  I'm a codeowner now but it still needs approval from one of you (which makes sense I guess, I can't approve my own PR)

**Review by wuisawesome — APPROVED** [review]
code owner files look fine

**Pull request merged**
